encryption-cbc.go

```diff
@@ -0,0 +1,258 @@
+/*
+ * Minio Go Library for Amazon S3 Compatible Cloud Storage (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package minio
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"io"
+	"net/http"
+)
+
+const (
+	AmzHeaderIV      = "X-Amz-Meta-X-Amz-Iv"
+	AmzHeaderKey     = "X-Amz-Meta-X-Amz-Key"
+	AmzHeaderMatDesc = "X-Amz-Meta-X-Amz-Matdesc"
+)
+
+// Crypt mode - encryption or decryption
+type cryptMode int
+
+const (
+	encryptMode cryptMode = iota
+	decryptMode
+)
+
+// Encrypt/Decrypt data using AES CBC algorithm
+type CBCSecureData struct {
+	// Stream to encrypt/decrypt
+	stream io.Reader
+
+	err error
+	eof bool
+
+	srcBuf *bytes.Buffer
+	dstBuf *bytes.Buffer
+
+	encryptionKey EncryptionKey
+	cryptedKey    []byte
+	contentKey    []byte
+	iv            []byte
+	matDesc       []byte
+
+	cryptMode cryptMode
+	blockMode cipher.BlockMode
+}
+
+func NewCBCSecureData(key EncryptionKey) *CBCSecureData {
+	return &CBCSecureData{
+		srcBuf:        bytes.NewBuffer([]byte{}),
```

Comment:
> Validate if key is valid..

The user is obliged to give something which satisfies EncryptionKey interface, I don't think we need any validation here.